### PR TITLE
Refactor Msf::Exploit::Deprecated and assorted fixes

### LINKED ISSUE #12026: Reloading a module alias fails, while reloading the OG module still works
```
msf5 > use auxiliary/scanner/smb/login
msf5 auxiliary(scanner/smb/login) > reload
[*] Reloading module...
[-] Failed to reload module:
msf5 auxiliary(scanner/smb/login) > use auxiliary/scanner/smb/smb_login
msf5 auxiliary(scanner/smb/smb_login) > reload
[*] Reloading module...
msf5 auxiliary(scanner/smb/smb_login) >
```

#11941

### PR DESCRIPTION
**Pull request by acammack-r7**
Adds a generic way of adding error and warning messages to modules to be displayed to users in `Msf::Module::Alert`. Used  twice in this PR, hopefully we can standardize some other error messages (like external module incompatibilities) to this mechanism.

Module reloads also now much faster as we avoid the slow module instance creation case of not providing a type to the module manager.

Fixes #12026 
Supersedes #12027 

Verification
========
#### New deprecation method:
- [x] `./msfconsole`
- [x] `use exploit/linux/http/ubiquiti_airos_file_upload`
- [x] Verify the module loads without error
- [x] `use exploit/linux/ssh/ubiquiti_airos_file_upload`
- [x] Verify the module loads and is `exploit/linux/http/ubiquiti_airos_file_upload` (but is still called `exploit/linux/ssh/ubiquiti_airos_file_upload` in the prompt) and displays the deprecation message

#### Alias reload fix:
- [x] `reload`
- [x] Verify the module reloads as `exploit/linux/ssh/ubiquiti_airos_file_upload`
- [x] `edit` the module and remove the `moved_from` call
- [x] `reload`
- [x] Verify the module reloads as `exploit/linux/http/ubiquiti_airos_file_upload` and prints a warning
- [x] `use exploit/linux/ssh/ubiquiti_airos_file_upload`
- [x] Verify the alias no longer works
- [x] `use exploit/linux/http/ubiquiti_airos_file_upload`
- [x] Verify the module loads and does not display a warning

### CODE REVIEW

**Comment by acammack-r7** [comment]
Looks like some of the tests relied on the `setup` jankiness, I'll address those.

File: modules/exploits/linux/http/ubiquiti_airos_file_upload.rb [hunk 1]
```diff
@@ -13,6 +13,9 @@ class MetasploitModule < Msf::Exploit::Remote
 
   include Msf::Exploit::Remote::HttpClient
   include Msf::Exploit::Remote::SSH
+  include Msf::Exploit::Deprecated
+
+  moved_from 'exploit/linux/ssh/ubiquiti_airos_file_upload'
```
Comment (wvu-r7) [inline modules/exploits/linux/http/ubiquiti_airos_file_upload.rb]: Love it!

File: lib/msf/core/modules/loader/base.rb [hunk 2]
```diff
@@ -283,14 +283,20 @@ def reload_module(original_metasploit_class_or_instance)
...
 
-    dlog("Reloading module #{module_reference_name}...", 'core')
+    dlog("Reloading module #{module_fullname}...", 'core')
 
     if load_module(parent_path, type, module_reference_name, :force => true, :reload => true)
-      # Create a new instance of the module
-      reloaded_module_instance = module_manager.create(module_reference_name)
+      # Create a new instance of the module, using the alias if one was used
+      reloaded_module_instance = module_manager.create(module_used_name || module_fullname)
+      if !reloaded_module_instance && module_fullname != module_used_name
```
Comment (wvu-r7) [inline lib/msf/core/modules/loader/base.rb]: Thanks for not using `unless`.

File: lib/msf/core/module/alert.rb [hunk 3]
```diff
@@ -0,0 +1,219 @@
+module Msf::Module::Alert
```
Comment (wvu-r7) [inline lib/msf/core/module/alert.rb]: This will be useful!

**Review by wvu — APPROVED** [review]
These look like good changes. I'll give it a test!

**Comment by wvu-r7** [comment]
Reminder to self: update #12133 when this lands.

**Comment by wvu-r7** [comment]
@acammack-r7: Travis failures may be related.

**Comment by wvu-r7** [comment]
Yeah, the `setup`/`setup` stuff is a little janky. Let's make sure it's ironclad before this goes in.

**Comment by acammack-r7** [comment]
If I can't get it figured out today I'll just revert it and tackle it later. From the standpoint of the PR it only means that some modules will have one less chance of displaying alerts to the user.

**Comment by acammack-r7** [comment]
@wvu-r7 The `super` change ended up being larger than what I wanted to take on so I left it out when I rebased to get some other specs playing nice.

**Comment by wvu-r7** [comment]
All good, better we leave it out for now if it's causing issues.

**Comment by wvu-r7** [comment]
@acammack-r7: Oh boy. Where I left off in the verification steps:

```
msf5 exploit(linux/http/ubiquiti_airos_file_upload) > use exploit/linux/ssh/ubiquiti_airos_file_upload
[-] Error while running command use: stack level too deep
msf5 exploit(linux/http/ubiquiti_airos_file_upload) >
```

**Comment by acammack-r7** [comment]
Hmm, that would be possible if the alias table isn't getting reset right, though on my box I cannot reproduce:
```
msf5 exploit(linux/ssh/ubiquiti_airos_file_upload) > reload
[*] Reloading module...

[!] Alias exploit/linux/ssh/ubiquiti_airos_file_upload no longer available after reloading, using exploit/linux/http/ubiquiti_airos_file_upload
msf5 exploit(linux/http/ubiquiti_airos_file_upload) > use exploit/linux/ssh/ubiquiti_airos_file_upload
[-] No results from search
[-] Failed to load module: exploit/linux/ssh/ubiquiti_airos_file_upload
```

**Comment by wvu-r7** [comment]
Much better.

```
msf5 exploit(linux/http/ubiquiti_airos_file_upload) > use exploit/linux/ssh/ubiquiti_airos_file_upload
[-] No results from search
[-] Failed to load module: exploit/linux/ssh/ubiquiti_airos_file_upload
msf5 exploit(linux/http/ubiquiti_airos_file_upload) >
```

**Comment by wvu-r7** [comment]
# Release Notes

This overhauls module deprecation by using module aliases, refactoring the deprecation mixin in the process. The new deprecation procedure allows a deprecated module to be deleted, transparently redirecting the user to the new module.

**Comment by wvu-r7** [comment]
Copied here for posterity:

> wvu  1 minute ago
> We may want to discuss a timeline for removing the `moved_from` line, but it's not essential
> wvu  1 minute ago
> Since the old code is gone already
> wvu  < 1 minute ago
> There is no fear of users using the deprecated module and filing bugs against it

**Comment by acammack-r7** [comment]
Whoops, some of the stuff here doesn't backport cleanly. I'll cherry-pick and open a PR to `4.x`.

**Comment by wvu-r7** [comment]
Ah, yeah, `evasion.rb` snuck in there.

**Pull request merged**